### ISSUE TITLE
`pathOr`: Allow paths that don't exist

### ISSUE THREAD

**Issue by dartess** [remeda/remeda#355]
Docs says:

```
R.pathOr({ x: 10 }, ['y'], 2) // 2
```

But it's not. Actually this won't work at all:

```
No overload matches this call.
  Overload 1 of 6, '(object: { x: number; }, path: readonly ["x"], defaultValue: number): number', gave the following error.
    Type '"y"' is not assignable to type '"x"'.
  Overload 2 of 6, '(object: { x: number; }, path: readonly ["x", "toString" | "toFixed" | "toExponential" | "toPrecision" | "valueOf" | "toLocaleString"], defaultValue: PathValue2<{ x: number; }, "x", "toString" | "toFixed" | "toExponential" | "toPrecision" | "valueOf" | "toLocaleString">): PathValue2<...>', gave the following error.
    Type '"y"' is not assignable to type '"x"'.
  Overload 3 of 6, '(object: { x: number; }, path: readonly ["x", "toString" | "toFixed" | "toExponential" | "toPrecision" | "valueOf" | "toLocaleString", never], defaultValue: never): never', gave the following error.
    Type '"y"' is not assignable to type '"x"'.ts(2769)
```

If you're checking paths at the type level, what's the point of the `Or` in a function name? If the path definitely exists and is type-checked, it should be `R.path( ... ) ?? defaultValue`. But if we allow any path to be passed, then the path should be just a `string[]`.

**Comment by dartess** [comment]
The defaultValue type is also broken. why is it tied to an object? because of this, I can't explicitly get undefined when the property hasn't been found. Because of this, I can't branch "the value in the path exists" and "the value in the path does not exist".

**Comment by eranhirsch** [comment]
I actually think the example is wrong because the case where you'd want to access prop "y" of an object that doesn't contain (at the type level) that prop makes sense. It's the same logic that we use in `prop` (which is actually just pathOr with a single prop). We do this because we want to protect users at compile time that the prop they are using is wrong, it's the advantage of using a well-typed utility library over more broad types that allow more cases but fail de-facto in runtime, often in production.

Having said all that, more broadly, I don't see the value in this utility, with modern eslint rules, prettier formatting, it's just easier to either use dot notation (e.g. `myObj?.myProp.myOtherprop?.myThirdProp.myLastProp ?? fallback`) or destructuring (e.g. `map(({ myProp: { myOtherProp: { myThirdProp: { myLastProp = fallback } } } } => myLastProp)`).

If you want to send a PR to fix the example I'd be happy to approve it, but I wouldn't change the typing here. If you feel the typing is generally bad, let's consider dropping this utility.

cc: @TkDodo, wdyt?

**Comment by dartess** [comment]
Okay, I think I get your point, but you don't seem to get mine.

1. There is no reason to type `defaultValue` by the path. Instead, you can return the union of the current type (by passed path) and `defaultValue`. This will allow for a wider use for `Or` in the name and will not mess up the typing in any way. I would also be happy to just use the `??` operator, but I can’t - now I must to pass `defaultValue` and I can’t pass `null` or `undefined` there (and I want to pass it there).

2. I love strongly typed code. But this is not always possible at all. I use this method in a utility to compare a typed object with a known structure and a JSON file. And if, when collecting the keys of a typed object, I could still get some types, but JSON is JSON. Also, I collect the keys as strings for ease of processing, and I use `stringToPath` to get the values in `pathOr`. Thus, I got a lot of pain from typing, and smeared the code with casts and `any`. It is not possible to use "optional chaining" here. This helper is needed here.

Your answer gave me the following idea: often your methods have two variations, regular and "strict". So, the current typing looks like `strict`, but I need a simple one. I think it is possible to transfer the current implementation to the `strict`. And in a simple way, allow you to pass a wider type for the `path` and for the `defaultValue`. If the path extends one of the paths in object, the return type will correspond to the union of the current return type and the `defaultValue` type. If not (for example, it's just an array of strings), the return type will match the union of all types in the object and the `defaultValue` type. Yes, it will be a change in typing, but it doesn't look like a breaking change for current users.

**Comment by dartess** [comment]
Right now this utility is useful for me. It works great, but only thanks to `@ts-ignore`. I literally don't need to have types here (so I don't need a `zod`), script should just get the difference between the keys and values of two objects (deeply).

In any case, I also consider it absolutely useless in the current strict version and is easily replaced by `obj?.[a]?.[b] ?? defaultValue`. If you don't consider lightening types, whether you keep the current version or remove the utility, I will need to rewrite my script.

**Comment by eranhirsch** [comment]
Regardless, because this is already out in the wild and being used by people we can only add a nonstrict variant so we don't break people's code. We can't move the typing here to a strict variant

**Comment by TkDodo** [comment]
I think the biggest advantage of our functions over optional chaining combined with nullish coalesce is that it can be used in pipes. I'm not sure if `pathOr` is useful there. If its not I think we should drop the function for v2

**Comment by dartess** [comment]
Please explain to me how can I use optional chaining when the key path comes to me from the backend in the string format lie `"foo.bar.baz"`?

**Comment by dartess** [comment]
> I think you need to split anyways because pathOr doesn't allow dot notation?

_I just can use `R.stringToPath` ;) this helper looks like a good match for `pathOr`_

In any case, in order not to slow down the migration, I completely rewrote all the parts where `pathOr` was used before. By the way, the new code looks better and contains more strict and explicit types. Deleting this helper may motivate other developers to make their code a little better.

So, I am no longer interested in this issue. It looks like it can be closed.

**Comment by TkDodo** [comment]
you're right, `stringToPath` was added to help with `pathOr`. I think if we decide to remove `pathOr`, we'll also remove `stringToPath`.

**Comment by eranhirsch** [comment]
We've made several major changes to `prop`, `defaultTo`, and `stringToPath`, and deprecated `pathOr`. With these recent changes, the core issues discussed here are resolved. Other things discussed here about relaxing the typing go against how we design our APIs, preferring to surface compile-time errors over allowing users more flexibility in what they can do with our functions.

If there are any remaining issues, they would be better served as a new issue.

See: #1145, #1150, #1158, #1159
